### PR TITLE
[O2B-1359] Display of synchronous flags at Runs Per LHC Period view

### PR DESCRIPTION
**Pull request by xsalonx**
#### I have a JIRA ticket
- [x] branch and/or PR name(s) include(s) JIRA ID
- [x] issue has "Fix version" assigned
- [x] issue "Status" is set to "In review"
- [x] PR labels are selected

Notable changes for users:
- Added display of synchronous QC flags

Notable changes for developers:
- NA

Changes made to the database:
- NA

### CODE REVIEW

**Comment by xsalonx** [comment]
To be merged after https://github.com/AliceO2Group/Bookkeeping/pull/1719

File: lib/public/views/Runs/RunPerPeriod/RunsPerLhcPeriodOverviewModel.js [hunk 1]
```diff
@@ -63,11 +63,93 @@ export class RunsPerLhcPeriodOverviewModel extends RunsOverviewModel {
...
+    async _fetchSynchronousQcSummary() {
+        if (this._lhcPeriodName) {
+            this.currentPanelData = RemoteData.loading();
+            this.notify();
+            try {
+                const { data: [lhcPeriod] } = await jsonFetch(`/api/lhcPeriodsStatistics?filter[names][]=${this._lhcPeriodName}`);
+                if (!lhcPeriod) {
+                    this.currentPanelData = RemoteData.failure([{ title: `Cannot find LHC period with name '${this._lhcPeriodName}'` }]);
+                    this.notify();
+                    return;
```
Comment (martinboulais) [inline lib/public/views/Runs/RunPerPeriod/RunsPerLhcPeriodOverviewModel.js:135]: I think that you should use else here, to benefit from the global notify? I think the code would be a bit more uniform. As you prefer.

File: lib/public/views/Runs/format/formatDetectorQuality.js [hunk 2]
```diff
@@ -10,17 +10,31 @@
  * granted to it by virtue of its status as an Intergovernmental Organization
  * or submit itself to any jurisdiction.
  */
-import { h, iconBan, switchCase } from '/js/src/index.js';
+import { getValueFromCssVar, isContrastBlack } from '../../../components/common/badge.js';
+import { h, iconBan } from '/js/src/index.js';
+
+const OnlineGeneralQualityColors = {
```
Comment (martinboulais) [inline lib/public/views/Runs/format/formatDetectorQuality.js:16]: Could you create a generic `Colors` enum? And in the same file export a function `getContrastColor` that takes a color, and either directly return its contrast if it's raw color or compute it using CSS if it's a CSS var.
The objective is that the CSS leak that we are doing is contained in only one file.

File: test/public/defaults.js [hunk 3]
```diff
@@ -145,17 +145,17 @@ module.exports.waitForTimeout = waitForTimeout;
  *
  * @param {puppeteer.Page} page - The puppeteer page where the table is located.
  * @param {number} expectedSize - The expected number of table rows, excluding rows marked as loading or empty.
+ * @param {string} [operator = '==='] comparison operator
  * @return {Promise<void>} Resolves once the expected number of rows is met, or the timeout is reached.
  */
-const waitForTableToLength = async (page, expectedSize) => {
+const waitForTableLength = async (page, expectedSize, operator = '===') => {
```
Comment (martinboulais) [inline test/public/defaults.js:151]: No. In tests you must know exactly what to expect.
`` makes the code harder to understand, have no autocomplete... I strongly disagree with this.
Also, you should refrain from changing the behavior of functions. Function named waitForTableToLength should do what it means (if you need to rename it it's even more obvious that you should not do it), if you need to extend the behavior you should simply create a new one.

File: test/public/runs/runsPerLhcPeriod.overview.test.js [hunk 4]
```diff
@@ -73,18 +75,13 @@ module.exports = () => {
...
-        // We expect the page to return the correct status code, making sure the server is running properly
         expect(response.status()).to.equal(200);
 
-        // We expect the page to return the correct title, making sure there isn't another server running on this port
         const title = await page.title();
         expect(title).to.equal('AliceO2 Bookkeeping');
     });
 
     it('shows correct datatypes in respective columns', async () => {
-        await goToPage(page, 'runs-per-lhc-period', { queryParameters: { lhcPeriodName: 'LHC22a' } });
```
Comment (martinboulais) [inline test/public/runs/runsPerLhcPeriod.overview.test.js:85]: Very nice 👍🏻

File: test/public/runs/runsPerLhcPeriod.overview.test.js [hunk 5]
```diff
@@ -178,24 +181,15 @@ module.exports = () => {
...
+        const expectedRunNumber = await getInnerText(await page.waitForSelector('tbody tr:first-of-type a'));
+        await waitForNavigation(page, () => pressElement(page, 'tbody tr:first-of-type a'));
+        expectUrlParams(page, { page: 'run-detail', runNumber: expectedRunNumber });
     });
 
     const EXPORT_RUNS_TRIGGER_SELECTOR = '#export-runs-trigger';
 
     it('should successfully export all runs per lhc Period', async () => {
-        await goToPage(page, 'runs-per-lhc-period', { queryParameters: { lhcPeriodName: 'LHC22a' } });
+        await waitForNavigation(page, () => page.goBack());
```
Comment (martinboulais) [inline test/public/runs/runsPerLhcPeriod.overview.test.js:192]: You are directly adding dependency on the test before here, you should avoid this.

File: lib/public/views/Runs/ActiveColumns/runDetectorsSyncQcActiveColumns.js [hunk 6]
```diff
@@ -10,23 +10,36 @@
...
+    { profiles, qcSummary } = {},
+) => Object.fromEntries(detectors?.map(({ name: detectorName, id: detectorId }) => [
     detectorName, {
         name: detectorName.toUpperCase(),
         visible: true,
-        format: (_, run) => formatDetectorQuality(run.detectorsQualities.find(({ name }) => name === detectorName)?.quality),
+        format: (_, { runNumber, detectorsQualities }) => {
+            const detectorWasActiveDuringRun = Boolean(detectorsQualities.find(({ name }) => name === detectorName));
+            if (!detectorWasActiveDuringRun) {
+                return null;
```
[the remark this answers is not in the document]
Comment (xsalonx) [reply lib/public/views/Runs/ActiveColumns/runDetectorsSyncQcActiveColumns.js:35]: use null everywhere

File: lib/public/views/Runs/RunPerPeriod/RunsPerLhcPeriodOverviewModel.js (hunk 1 above)
Comment (xsalonx) [reply lib/public/views/Runs/RunPerPeriod/RunsPerLhcPeriodOverviewModel.js:135]: right, applied

File: test/public/defaults.js (hunk 3 above)
Comment (xsalonx) [reply test/public/defaults.js:151]: applied

File: lib/public/views/Runs/format/formatDetectorQuality.js (hunk 2 above)
Comment (xsalonx) [reply lib/public/views/Runs/format/formatDetectorQuality.js:16]: applied

Comment (xsalonx) [reply lib/public/views/Runs/format/formatDetectorQuality.js:16]: done

File: test/public/runs/runsPerLhcPeriod.overview.test.js (hunk 5 above)
Comment (xsalonx) [reply test/public/runs/runsPerLhcPeriod.overview.test.js:192]: applied

File: lib/public/components/common/colors.js [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    return red * 0.299 + green * 0.587 + blue * 0.114 > 186;
+};
+
+/**
+ * Get value of CSS variable
+ *
+ * @param {string} variableName variable name
+ * @return {string} value
+ */
+export const getValueFromCssVar = (variableName) => getComputedStyle(document.documentElement).getPropertyValue(variableName);
```
Comment (martinboulais) [inline lib/public/components/common/colors.js:67]: I think that this one and the `isContrastBlack` should not be exported.

**Review by martinboulais — CHANGES_REQUESTED** [review]

File: lib/public/components/common/colors.js (hunk 7 above)
Comment (xsalonx) [reply lib/public/components/common/colors.js:67]: applied

**Review by martinboulais — DISMISSED** [review]

**Review by martinboulais — APPROVED** [review]

**Pull request merged**